### PR TITLE
Implement the hipOccupancyMaxPotentialBlockSize function

### PR DESCRIPTION
**Pull request by wkwchau**
This version supports both CO v2 and v3.

### CODE REVIEW

**Comment by whchung** [comment]
Nice!  + @deven-amd / @jerryyin  for awareness.

File: src/hip_module.cpp [hunk 1]
```diff
@@ -652,3 +652,145 @@ hipError_t hipModuleGetTexRef(textureReference** texRef, hipModule_t hmod, const
...
+
+    auto ctx = ihipGetTlsDefaultCtx();
+    hipError_t ret = hipSuccess;
+
+    if (ctx == nullptr) {
+        ret = hipErrorInvalidDevice;
+    }
+
+    hipDeviceProp_t prop{};
+    hipGetDeviceProperties(&prop, ihipGetTlsDefaultCtx()->getDevice()->_deviceId);
```
Comment (mangupta) [inline src/hip_module.cpp]: Replace with ihipGetDeviceProperties() otherwise it will mess up the HIP trace.

Comment (wkwchau) [reply src/hip_module.cpp]: Fixed.

**Comment by gargrahul** [comment]
@wkwchau Could you add a test too?

**Comment by gargrahul** [comment]
@wkwchau It needs ../../test_common.cpp.

Also isn't it possible to use this API without using Module APIs calls?

**Comment by gargrahul** [comment]
@mangupta I have disabled the hipOccupancyMaxPotentialBlockSize dtest on NVCC for now. We need to discuss how to support both runtime and driver APIs support with single HIP API.

File: include/hip/hcc_detail/hip_runtime.h [hunk 2]
```diff
@@ -279,6 +279,8 @@ extern hipStream_t ihipPreLaunchKernel(hipStream_t stream, size_t grid, size_t b
                                        grid_launch_parm* lp, const char* kernelNameStr);
 extern void ihipPostLaunchKernel(const char* kernelName, hipStream_t stream, grid_launch_parm& lp);
 
+extern hipError_t ihipGetDeviceProperties(hipDeviceProp_t* props, int device);
```
Comment (mangupta) [inline include/hip/hcc_detail/hip_runtime.h]: Why do we need this here? This should go to src/hip_module.cpp or src/hip_hcc_internal.h.

Comment (wkwchau) [reply include/hip/hcc_detail/hip_runtime.h]: Fixed by moving extern declaration of ihipGetDeviceProperties to hip_module.cpp.

File: src/hip_module.cpp [hunk 3]
```diff
@@ -652,3 +654,147 @@ hipError_t hipModuleGetTexRef(textureReference** texRef, hipModule_t hmod, const
...
+        }
+
+        size_t maxWavesLDSLimited = min(wavefrontsLDS, maxWavefrontsPerCU);
+
+        size_t activeWavefronts = 0;
+        size_t tmp_min = (size_t)min(maxWavesLDSLimited, maxWavesWGLimited);
+        tmp_min = min(maxWavesSGPRSLimited, tmp_min);
+        activeWavefronts = min(maxWavesVGPRSLimited, tmp_min);
+
+        if (maxActivWaves < activeWavefronts) {
```
Comment (whchung) [inline src/hip_module.cpp]: @wkwchau shall this line be `<=` instead of `<`?

**Pull request merged**